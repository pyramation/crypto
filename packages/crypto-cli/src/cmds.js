
import _createWallet from './commands/create-wallet';
import _customWallet from './commands/custom-wallet';
import _mnemonicWallet from './commands/mnemonic-wallet';
import _newCoinForWalletGeneratorFromKeys from './commands/new-coin-for-wallet-generator-from-keys';
import _saltDecrypt from './commands/salt-decrypt';
import _saltEncyrpt from './commands/salt-encyrpt';
import _saltGenerate from './commands/salt-generate';
import _signMessage from './commands/sign-message';
import _verifyKey from './commands/verify-key';
import _verifyMessage from './commands/verify-message';
import _walletFromMnemonic from './commands/wallet-from-mnemonic';
const Commands = {};
Commands['createWallet'] = _createWallet;
Commands['customWallet'] = _customWallet;
Commands['mnemonicWallet'] = _mnemonicWallet;
Commands['newCoinForWalletGeneratorFromKeys'] = _newCoinForWalletGeneratorFromKeys;
Commands['saltDecrypt'] = _saltDecrypt;
Commands['saltEncyrpt'] = _saltEncyrpt;
Commands['saltGenerate'] = _saltGenerate;
Commands['signMessage'] = _signMessage;
Commands['verifyKey'] = _verifyKey;
Commands['verifyMessage'] = _verifyMessage;
Commands['walletFromMnemonic'] = _walletFromMnemonic;

  export { Commands }; 

  