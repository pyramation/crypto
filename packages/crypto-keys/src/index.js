import allCoins from '@pyramation/crypto-coins';
import * as bitcoin from 'bitcoinjs-lib';
import * as bs58check from 'bs58check';

export const getNetworkFromHashAndWifVersion = (pubKeyHash, wif) => {
  return {
    messagePrefix: '\x19MyCoin Signed Message:\n',
    bip32: {
      public: 0x0, // TODO DONT USE
      private: 0x0 // TODO DONT USE
    },
    pubKeyHash,
    scriptHash: 0x0, // TODO DONT USE
    wif
  };
};

export const getChars = (pubKeyHash, wif) => {
  const network = getNetworkFromHashAndWifVersion(pubKeyHash, wif);

  const cmp = {};
  for (let i = 0; i < 250; i++) {
    const pair = bitcoin.ECPair.makeRandom({ network, compressed: true });
    cmp[pair.toWIF()[0]] = 1;
  }

  const uncmp = {};
  for (let i = 0; i < 250; i++) {
    const pair = bitcoin.ECPair.makeRandom({ network, compressed: false });
    uncmp[pair.toWIF()[0]] = 1;
  }

  return {
    compressed: Object.keys(cmp).sort(),
    uncompressed: Object.keys(uncmp).sort()
  };
};

export const pubToNetworkPrefix = (publicAddr) => {
  return bitcoin.address.fromBase58Check(publicAddr).version.toString(16);
};

export const privToPrivPrefix = (privateKey) => {
  return bs58check.decode(privateKey)[0].toString(16);
};

export const networkInfoFromKeys = function (pub, priv) {
  const networkVersion = pubToNetworkPrefix(pub);
  const privateKeyPrefix = privToPrivPrefix(priv);

  let { compressed, uncompressed } = getChars(
    parseInt(networkVersion, 16),
    parseInt(privateKeyPrefix, 16)
  );

  if (uncompressed.length > 1) uncompressed = `[${uncompressed.join('')}]`;
  else uncompressed = uncompressed[0];
  if (compressed.length > 1) compressed = `[${compressed.join('')}]`;
  else compressed = compressed[0];

  return {
    networkVersion,
    privateKeyPrefix,
    uncompressed,
    compressed
  };
};

export const infoForWalletNet = function (pub, priv) {
  const {
    networkVersion,
    privateKeyPrefix,
    uncompressed,
    compressed
  } = networkInfoFromKeys(pub, priv);

  const similar = allCoins
    .filter((c) => {
      if (
        c[1] === parseInt(networkVersion, 16) &&
        c[2] === parseInt(privateKeyPrefix, 16)
      ) {
        return true;
      }
    })
    .map((c) => c[0]);

  return {
    networkVersion,
    privateKeyPrefix,
    compressed,
    uncompressed,
    similar,
    message: `array to add to WalletGenerator.net:
<nameOfCoin>
0x${networkVersion}
0x${privateKeyPrefix}
${uncompressed}
${compressed}
<donationAddr>
    `
  };
};

export const networkFromCurrencyGeneratorArray = (array) => {
  return getNetworkFromHashAndWifVersion(array[1], array[2]);
};

export const newWallet = (network) => {
  const pair = bitcoin.ECPair.makeRandom({ network });
  return {
    public: getAddress(pair, network),
    private: pair.toWIF(),
    pair
  };
};

export const getAddress = (pair, network) => {
  return bitcoin.payments.p2pkh({ pubkey: pair.publicKey, network }).address;
};

export const verifyPrivate = (privateKey, network) => {
  const pair = bitcoin.ECPair.fromWIF(privateKey, network);
  return getAddress(pair, network);
};
