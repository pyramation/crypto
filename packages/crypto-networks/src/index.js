export default {
  // https://github.com/bitcoin/bitcoin/blob/313a081b907bf0a5b56af99ec2d42814ef0638b0/src/chainparams.cpp#L126-L132
  BTC: {
    messagePrefix: '\x19Bitcoin Signed Message:\n',
    bech32: 'bc', // bech32_hrp
    bip32: {
      public: 0x0488b21e, // base58Prefixes[EXT_PUBLIC_KEY]
      private: 0x0488ade4 // base58Prefixes[EXT_SECRET_KEY]
    },
    pubKeyHash: 0x00, // base58Prefixes[PUBKEY_ADDRESS] (0)
    scriptHash: 0x05, // base58Prefixes[SCRIPT_ADDRESS] (5)
    wif: 0x80 // base58Prefixes[SECRET_KEY] (128)
  },
  BITG: {
    // https://github.com/bitgreen/bitgreen/blob/13982be25d11a75d012b2778ac7c85bc11726d78/src/chainparams.cpp#L227-L233  BitGreen: {
    // https://github.com/bitgreen/bitgreen/blob/36e2b1fb3602eda482383b2fd1583a2a7aca9d4d/src/util/validation.cpp#L20
    messagePrefix: '\x19BitGreen Signed Message:\n',
    bech32: 'bg', // bech32_hrp
    bip32: {
      public: 0x0488b21e, // base58Prefixes[EXT_PUBLIC_KEY]
      private: 0x0488ade4 // base58Prefixes[EXT_SECRET_KEY]
    },
    pubKeyHash: 0x26, // base58Prefixes[PUBKEY_ADDRESS] (38)
    scriptHash: 0x06, // base58Prefixes[SCRIPT_ADDRESS] (6)
    wif: 0x2e // base58Prefixes[SECRET_KEY] (46)
  }
};
