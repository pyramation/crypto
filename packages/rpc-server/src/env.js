import { cleanEnv, bool, str, port, url } from 'envalid';

module.exports = cleanEnv(
  process.env,
  {
    SERVER_PORT: port({ default: 6969 }),
    SERVER_HOST: str({ default: 'localhost' }),
    DAEMON_URL: url()
  },
  { dotEnvPath: null }
);
