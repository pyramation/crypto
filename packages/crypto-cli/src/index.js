import inquirer from 'inquirer';
import { Commands as commands } from './cmds';

const keys = Object.keys(commands);

inquirer
  .prompt([
    {
      type: 'list',
      name: 'cmd',
      message: 'what do you want to do?',
      choices: keys.map((key) => ({
        name: key,
        value: key
      }))
    }
  ])
  .then((result) => {
    commands[result.cmd]();
  });
