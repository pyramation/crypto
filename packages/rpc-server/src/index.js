
import express from 'express';
import request from 'request';
import bodyParser from 'body-parser';
import env from './env';

const app = express();
app.use(bodyParser.json());

const url = /^http/.test(env.DAEMON_URL)
    ? env.DAEMON_URL
    : `http://${env.DAEMON_URL}`;

const makereq = (method, params = []) => {
    const body = { jsonrpc: '1.0', id: 'curltext', method, params };
    return new Promise((resolve, reject) => {
        request.post(
            {
                headers: {
                    'content-type': 'text/plain;'
                },
                url,
                body: JSON.stringify(body)
            },
            function (error, response, data) {
                if (error) {
                    return reject(error);
                }
                try {
                    const json = JSON.parse(data);
                    if (json.error) {
                        return reject(json.error);
                    }
                    return resolve(json.result);
                } catch (e) {
                    return reject(e);
                }
            }
        );
    });
};

const str = (i) => i;
const int = (i) => parseInt(i, 10);

const COMMANDS = [
    ['getblockhash', int],
    ['getblock', str],
    ['getrawtransaction', str],
    ['decoderawtransaction', str],
    ['sendrawtransaction', str],
    ['getblockcount'],
    ['getbestblockhash'],
    ['getconnectioncount'],
    ['getdifficulty'],
    ['getblockchaininfo'],
    ['getmininginfo'],
    ['getpeerinfo'],
    ['getrawmempool'],
];
const DIG = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'];
const makeUrl = (method, args) => {
    if (!args.length) return `/${method}`;
    return `/${method}/` + args.map((arg, i) => `:${DIG[i]}`).join('/');
}

const callReq = async (method, req, args) => {
    if (!args.length) return await makereq(method);
    const casted = args.map((cast, i)=>  cast(req.params[DIG[i]]));
    return await makereq(method, casted);
}

COMMANDS.forEach(([method, ...args]) => {
    app.get(makeUrl(method, args), async (req, res, next) => {
        try {
            const result = await callReq(method, req, args);
            res.status(200);
            res.json(result);
        } catch (e) {
            next(e);
        }
    });
});

app.use(async (error, req, res, next) => {
    return res.status(500).send({ message: error.message });
});

export default () => {
    app.listen(env.SERVER_PORT, env.SERVER_HOST, () =>
        console.log(`app listening: ${env.SERVER_HOST}:${env.SERVER_PORT}`)
    );
};