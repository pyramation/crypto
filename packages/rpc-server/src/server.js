import express from 'express';
import request from 'request';
import bodyParser from 'body-parser';
import env from './env';

const app = express();
app.use(bodyParser.json());

const url = /^http/.test(env.DAEMON_URL)
  ? env.DAEMON_URL
  : `http://${env.DAEMON_URL}`;

const makereq = (method, params = []) => {
  const body = { jsonrpc: '1.0', id: 'curltext', method, params };
  console.log(JSON.stringify(body));
  return new Promise((resolve, reject) => {
    request.post(
      {
        headers: {
          'content-type': 'text/plain;'
        },
        url,
        body: JSON.stringify(body)
      },
      function (error, response, data) {
        if (error) {
          return reject(error);
        }
        try {
          const json = JSON.parse(data);
          return resolve(json.result);
        } catch (e) {
          return reject(e);
        }
      }
    );
  });
};

const str = (i) => i;
const int = (i) => parseInt(i, 10);

// 1 param
[
  ['getblockhash', int],
  ['getblock', str],
  ['getrawtransaction', str],
  ['decoderawtransaction', str]
].forEach(([method, cast]) => {
  app.get(`/${method}/:first`, async (req, res, next) => {
    try {
      const result = await makereq(method, [cast(req.params.first)]);
      res.set('Content-Type', 'application/json');
      res.status(200).send(result);
    } catch (e) {
      next(e);
    }
  });
});

// 0 param
[
  'getblockcount',
  'getbestblockhash',
  'getconnectioncount',
  'getdifficulty',
  'getblockchaininfo',
  'getmininginfo',
  'getpeerinfo',
  'getrawmempool'
].forEach((method) => {
  app.get(`/${method}`, async (req, res, next) => {
    try {
      const result = await makereq(method);
      res.set('Content-Type', 'application/json');
      res.status(200).send(result);
    } catch (e) {
      next(e);
    }
  });
});

app.use(async (error, req, res, next) => {
  return res.status(500).send({ message: error.message });
});

app.listen(env.SERVER_PORT, env.SERVER_HOST, () =>
  console.log(`app listening: ${env.SERVER_HOST}:${env.SERVER_PORT}`)
);
