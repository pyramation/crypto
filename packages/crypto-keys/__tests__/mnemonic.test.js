import * as bip32 from 'bip32';
import * as bip39 from 'bip39';
import * as bitcoin from 'bitcoinjs-lib';
import * as bs58check from 'bs58check';
import * as wif from 'wif';

import { getAddress } from '../src';

// just treat it as a 4 byte hex, like uint32_t
// bip32 pub is {0x04, 0x88, 0xB2, 0x1E} => 0x0488B21E => 76067358
// bip32 priv is {0x04, 0x88, 0xAD, 0xE4} => 0x0488ADE4 => 76066276

// so ive just signed 'gday hows it going'
// public key GWhXzs3hCWNasa1jX5xaSZ6h3uDgc7Lb9H
// signature HwxUI5nHX63tuudKnNzdYTvir58oMo2/p8s44fEub1B/LiBexpj1c/66stsnShITqGePlV1/SqfIYRA4MUOPyvQ=

const network = {
  //   messagePrefix: 'BitGreen Signed Message:\n',
  messagePrefix: '\x19BitGreen Signed Message:\n',
  bech32: 'bg',
  bip32: {
    public: 0x0488b21e,
    private: 0x0488ade4
  },
  pubKeyHash: 0x26,
  scriptHash: 0x6,
  wif: 0x2e
};

describe('bip39', () => {
  it('verify mnemonic', () => {
    const mnemonic =
      'praise you muffin lion enable neck grocery crumble super myself license ghost';
    const seed = bip39.mnemonicToSeedSync(mnemonic);
    const node = bip32.fromSeed(seed, network);
    const strng = node.toBase58();
    const restored = bip32.fromBase58(strng, network);

    expect(getAddress(node, network)).toEqual(getAddress(restored, network)); // same public key
    expect(node.toWIF()).toEqual(restored.toWIF()); // same private key

    expect(mnemonic).toMatchSnapshot();
    expect(getAddress(node, network)).toMatchSnapshot();
    expect(node.toWIF()).toMatchSnapshot();
  });
  it('generate mnemonic', () => {
    const mnemonic = bip39.generateMnemonic();
    const seed = bip39.mnemonicToSeedSync(mnemonic);
    const node = bip32.fromSeed(seed, network);
    const strng = node.toBase58();
    const restored = bip32.fromBase58(strng, network);
    expect(getAddress(node, network)).toEqual(getAddress(restored, network)); // same public key
    expect(node.toWIF()).toEqual(restored.toWIF()); // same private key
  });
});
