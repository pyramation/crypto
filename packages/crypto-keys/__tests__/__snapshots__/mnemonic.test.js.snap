// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bip39 verify mnemonic 1`] = `"praise you muffin lion enable neck grocery crumble super myself license ghost"`;

exports[`bip39 verify mnemonic 2`] = `"GKVPivuTA5yh8MfrA1gfqLdjmDhutyjmZb"`;

exports[`bip39 verify mnemonic 3`] = `"7oiKJETU51vWjAAVWRTBqi1Q9yQ8feLD36tBo4aiPsHnygqrGw5U"`;
